File: src/engine/Cards.ts
import type { EntityStats } from './Entity';

export const Cards: Record<string, EntityStats> = {
    knight: {
        id: 'knight',
        name: 'Knight',
        type: 'troop',
        hp: 1666,
        damage: 202,
        hitSpeed: 1.2,
        speed: 1.0, // Medium
        range: 0.8, // Melee
        sightRange: 5.5,
        radius: 0.5,
        mass: 6,
        isAir: false,
        targetType: 'ground'
    },
    archers: {
        id: 'archers',
        name: 'Archers',
        type: 'troop',
        hp: 304,
        damage: 107,
        hitSpeed: 0.9,
        speed: 1.0, // Medium
        range: 5.0,
        sightRange: 5.5,
        radius: 0.5,
        mass: 3,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10,
        spawnCount: 2,
        spawnRadius: 0.8
    },
    giant: {
        id: 'giant',
        name: 'Giant',
        type: 'troop',
        hp: 4091,
        damage: 254,
        hitSpeed: 1.5,
        speed: 0.75, // Slow
        range: 0.8,
        sightRange: 5.5,
        radius: 0.75,
        mass: 18,
        isAir: false,
        targetType: 'building',
        isBuildingTargeter: true
    },
    pekka: {
        id: 'pekka',
        name: 'P.E.K.K.A',
        type: 'troop',
        hp: 3760,
        damage: 816,
        hitSpeed: 1.8,
        speed: 0.75, // Slow
        range: 0.8,
        sightRange: 5.5,
        radius: 0.8,
        mass: 18,
        isAir: false,
        targetType: 'ground'
    },
    musketeer: {
        id: 'musketeer',
        name: 'Musketeer',
        type: 'troop',
        hp: 720,
        damage: 218,
        hitSpeed: 1.1,
        speed: 1.0, // Medium
        range: 6.0,
        sightRange: 6.0,
        radius: 0.5,
        mass: 4,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10
    },
    hog_rider: {
        id: 'hog_rider',
        name: 'Hog Rider',
        type: 'troop',
        hp: 1696,
        damage: 318,
        hitSpeed: 1.6,
        speed: 2.0, // Very Fast
        range: 0.8,
        sightRange: 5.5,
        radius: 0.6,
        mass: 8,
        isAir: false,
        targetType: 'building',
        isBuildingTargeter: true,
        jumpsRiver: true
    },
    skeletons: {
        id: 'skeletons',
        name: 'Skeletons',
        type: 'troop',
        hp: 81,
        damage: 81,
        hitSpeed: 1.0,
        speed: 1.5, // Fast
        range: 0.8,
        sightRange: 5.5,
        radius: 0.4,
        mass: 1,
        isAir: false,
        targetType: 'ground',
        spawnCount: 3,
        spawnRadius: 0.5
    },
    elite_barbarians: {
        id: 'elite_barbarians',
        name: 'Elite Barbarians',
        type: 'troop',
        hp: 1341,
        damage: 382,
        hitSpeed: 1.4,
        speed: 2.0, // Very Fast
        range: 0.8,
        sightRange: 5.5,
        radius: 0.5,
        mass: 6,
        isAir: false,
        targetType: 'ground',
        spawnCount: 2,
        spawnRadius: 0.8
    },
    evo_knight: {
        id: 'evo_knight',
        name: 'Evo Knight',
        type: 'troop',
        hp: 1999,
        damage: 202,
        hitSpeed: 1.2,
        speed: 1.0, // Medium
        range: 0.8,
        sightRange: 5.5,
        radius: 0.5,
        mass: 6,
        isAir: false,
        targetType: 'ground',
        movingShieldDamageReduction: 0.6 // 60% shield while moving
    },
    evo_skeletons: {
        id: 'evo_skeletons',
        name: 'Evo Skeletons',
        type: 'troop',
        hp: 81,
        damage: 81,
        hitSpeed: 1.0,
        speed: 1.5, // Fast
        range: 0.8,
        sightRange: 5.5,
        radius: 0.4,
        mass: 1,
        isAir: false,
        targetType: 'ground',
        spawnCount: 3,
        spawnRadius: 0.5,
        spawnOnHitId: 'evo_skeleton_single',
        maxClones: 8
    },
    evo_skeleton_single: {
        id: 'evo_skeleton_single',
        name: 'Evo Skeleton',
        type: 'troop',
        hp: 81,
        damage: 81,
        hitSpeed: 1.0,
        speed: 1.5, // Fast
        range: 0.8,
        sightRange: 5.5,
        radius: 0.4,
        mass: 1,
        isAir: false,
        targetType: 'ground',
        spawnOnHitId: 'evo_skeleton_single',
        maxClones: 8
    },
    golden_knight: {
        id: 'golden_knight',
        name: 'Golden Knight',
        type: 'troop',
        hp: 1800,
        damage: 160,
        hitSpeed: 0.9,
        speed: 1.0, // Medium
        range: 1.2,
        sightRange: 5.5,
        radius: 0.5,
        mass: 5,
        isAir: false,
        targetType: 'ground',
        ability: {
            name: 'Dashing Dash',
            cooldown: 11,
            elixirCost: 1,
            effect: 'dashing_dash',
            radius: 5.0
        }
    },
    archer_queen: {
        id: 'archer_queen',
        name: 'Archer Queen',
        type: 'troop',
        hp: 1000,
        damage: 225,
        hitSpeed: 1.2,
        speed: 1.0, // Medium
        range: 5.0,
        sightRange: 5.0,
        radius: 0.5,
        mass: 3,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10,
        ability: {
            name: 'Cloaking Cape',
            cooldown: 15,
            elixirCost: 1,
            effect: 'cloaking_cape',
            duration: 3.0
        }
    },    elite_knight: {
        id: 'elite_knight',
        name: 'Hero Knight',
        type: 'troop',
        hp: 1666, // Base stats same as normal knight
        damage: 202,
        hitSpeed: 1.2,
        speed: 1.0, // Medium
        range: 1.2,
        sightRange: 5.5,
        radius: 0.5,
        mass: 3,
        isAir: false,
        targetType: 'ground',
        ability: {
            name: 'Triumphant Taunt',
            cooldown: 25,
            elixirCost: 2,
            effect: 'taunt_shield',
            duration: 5.0,
            radius: 6.5 // Taunt radius
        }
    },
    elite_musketeer: {
        id: 'elite_musketeer',
        name: 'Hero Musketeer',
        type: 'troop',
        hp: 883, // Exactly 883 HP per Season 77
        damage: 204, // Ranged attack damage
        hitSpeed: 1.3, // 1.3s hit speed
        speed: 1.0, // Medium
        range: 6.0,
        sightRange: 6.0,
        radius: 0.5,
        mass: 4,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10,
        ability: {
            name: 'Trusty Turret',
            cooldown: 15,
            elixirCost: 3,
            effect: 'spawn_turret'
        }
    },
    elite_turret: {
        id: 'elite_turret',
        name: 'Elite Turret',
        type: 'building',
        hp: 350,
        damage: 100,
        hitSpeed: 1.0,
        speed: 0,
        range: 3.5, // 2026 Feb balance change
        sightRange: 3.5,
        radius: 0.6,
        mass: Infinity,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10,
        lifetime: 10
    },
    elite_giant: {
        id: 'elite_giant',
        name: 'Elite Giant',
        type: 'troop',
        hp: 4091, // Exactly same as 11级 Giant
        damage: 254,
        hitSpeed: 1.5,
        speed: 0.5, // Slow
        range: 1.2,
        sightRange: 5.5,
        radius: 0.8,
        mass: 8,
        isAir: false,
        targetType: 'building',
        isBuildingTargeter: true,
        ability: {
            name: 'Throw Enemies',
            cooldown: 12,
            elixirCost: 1,
            effect: 'throw_enemies',
            radius: 3.0
        }
    },
    battle_healer: {
        id: 'battle_healer',
        name: 'Battle Healer',
        type: 'troop',
        hp: 1717,
        damage: 148,
        hitSpeed: 1.5,
        speed: 1.0, // Medium
        range: 0.8,
        sightRange: 5.5,
        radius: 0.6,
        mass: 6,
        isAir: false,
        targetType: 'ground'
    },
    princess_tower: {
        id: 'princess_tower',
        name: 'Princess Tower',
        type: 'tower',
        hp: 3052,
        damage: 109,
        hitSpeed: 0.8,
        speed: 0,
        range: 7.5,
        sightRange: 7.5,
        radius: 1.5,
        mass: Infinity,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10
    },
    king_tower: {
        id: 'king_tower',
        name: 'King Tower',
        type: 'tower',
        hp: 4824,
        damage: 109,
        hitSpeed: 1.0,
        speed: 0,
        range: 7.0,
        sightRange: 7.0,
        radius: 2.0,
        mass: Infinity,
        isAir: false,
        targetType: 'all',
        projectileSpeed: 10
    }
};

File: openroyale_sim/cards.py
from __future__ import annotations

from dataclasses import dataclass, replace
from math import inf
from typing import Literal

Team = Literal["blue", "red"]
TargetType = Literal["ground", "air", "building", "all"]
EntityType = Literal["troop", "building", "tower"]
AbilityEffect = Literal[
    "taunt_shield",
    "spawn_turret",
    "throw_enemies",
    "dashing_dash",
    "cloaking_cape",
    "summon_skeletons",
    "none",
]


@dataclass(frozen=True)
class AbilityStats:
    name: str
    cooldown: float
    elixir_cost: float
    effect: AbilityEffect
    duration: float = 0.0
    radius: float = 0.0


@dataclass(frozen=True)
class EntityStats:
    id: str
    name: str
    type: EntityType
    hp: float
    damage: float
    hit_speed: float
    load_time: float
    speed: float
    range: float
    sight_range: float
    radius: float
    mass: float
    target_type: TargetType
    is_air: bool
    elixir_cost: int = 0
    is_building_targeter: bool = False
    spawn_count: int = 1
    spawn_radius: float = 1.0
    jumps_river: bool = False
    projectile_speed: float | None = None
    projectile_asset: str | None = None
    projectile_trajectory: Literal["line", "parabola"] | None = None
    splash_radius: float | None = None
    moving_shield_damage_reduction: float | None = None
    spawn_on_hit_id: str | None = None
    max_clones: int | None = None
    ability: AbilityStats | None = None
    lifetime: float | None = None

    def with_overrides(self, **kwargs: object) -> "EntityStats":
        return replace(self, **kwargs)


CARDS: dict[str, EntityStats] = {
    "knight": EntityStats(
        id="knight",
        name="Knight",
        type="troop",
        hp=1666,
        damage=202,
        hit_speed=1.2,
        load_time=0.7,
        speed=1.0,
        range=0.8,
        sight_range=5.5,
        radius=0.5,
        mass=6,
        is_air=False,
        target_type="ground",
        elixir_cost=3,
    ),
    "archers": EntityStats(
        id="archers",
        name="Archers",
        type="troop",
        hp=304,
        damage=107,
        hit_speed=0.9,
        load_time=1.1,
        speed=1.0,
        range=5.0,
        sight_range=5.5,
        radius=0.5,
        mass=3,
        is_air=False,
        target_type="all",
        elixir_cost=3,
        projectile_speed=10,
        projectile_asset="projectile_arrow_basic",
        projectile_trajectory="parabola",
        spawn_count=2,
        spawn_radius=0.8,
    ),
    "giant": EntityStats(
        id="giant",
        name="Giant",
        type="troop",
        hp=4091,
        damage=254,
        hit_speed=1.5,
        load_time=1.0,
        speed=0.75,
        range=0.8,
        sight_range=5.5,
        radius=0.75,
        mass=18,
        is_air=False,
        target_type="building",
        elixir_cost=5,
        is_building_targeter=True,
    ),
    "pekka": EntityStats(
        id="pekka",
        name="P.E.K.K.A",
        type="troop",
        hp=3760,
        damage=816,
        hit_speed=1.8,
        load_time=1.3,
        speed=0.75,
        range=0.8,
        sight_range=5.5,
        radius=0.8,
        mass=18,
        is_air=False,
        target_type="ground",
        elixir_cost=7,
    ),
    "musketeer": EntityStats(
        id="musketeer",
        name="Musketeer",
        type="troop",
        hp=720,
        damage=218,
        hit_speed=1.1,
        load_time=0.6,
        speed=1.0,
        range=6.0,
        sight_range=6.0,
        radius=0.5,
        mass=4,
        is_air=False,
        target_type="all",
        elixir_cost=4,
        projectile_speed=10,
        projectile_asset="projectile_cannonball_small",
        projectile_trajectory="line",
    ),
    "hog_rider": EntityStats(
        id="hog_rider",
        name="Hog Rider",
        type="troop",
        hp=1696,
        damage=318,
        hit_speed=1.6,
        load_time=1.0,
        speed=2.0,
        range=0.8,
        sight_range=5.5,
        radius=0.6,
        mass=8,
        is_air=False,
        target_type="building",
        elixir_cost=4,
        is_building_targeter=True,
        jumps_river=True,
    ),
    "skeletons": EntityStats(
        id="skeletons",
        name="Skeletons",
        type="troop",
        hp=81,
        damage=81,
        hit_speed=1.0,
        load_time=0.5,
        speed=1.5,
        range=0.8,
        sight_range=5.5,
        radius=0.4,
        mass=1,
        is_air=False,
        target_type="ground",
        elixir_cost=1,
        spawn_count=3,
        spawn_radius=0.5,
    ),
    "barbarians": EntityStats(
        id="barbarians",
        name="Barbarians",
        type="troop",
        hp=670,
        damage=191,
        hit_speed=1.4,
        load_time=1.0,
        speed=1.0,
        range=0.8,
        sight_range=5.5,
        radius=0.5,
        mass=4,
        is_air=False,
        target_type="ground",
        elixir_cost=5,
        spawn_count=5,
        spawn_radius=1.0,
    ),
    "minions": EntityStats(
        id="minions",
        name="Minions",
        type="troop",
        hp=230,
        damage=105,
        hit_speed=1.0,
        load_time=1.0,
        speed=1.5,
        range=1.5,
        sight_range=5.5,
        radius=0.4,
        mass=2,
        is_air=True,
        target_type="all",
        elixir_cost=3,
        spawn_count=3,
        spawn_radius=0.8,
    ),
    "wizard": EntityStats(
        id="wizard",
        name="Wizard",
        type="troop",
        hp=720,
        damage=281,
        hit_speed=1.4,
        load_time=1.0,
        speed=1.0,
        range=5.5,
        sight_range=6.0,
        radius=0.5,
        mass=4,
        is_air=False,
        target_type="all",
        elixir_cost=5,
        projectile_speed=10,
        projectile_asset="fireball_projectile1",
        projectile_trajectory="line",
        splash_radius=1.5,
    ),
    "princess": EntityStats(
        id="princess",
        name="Princess",
        type="troop",
        hp=261,
        damage=169,
        hit_speed=3.0,
        load_time=1.0,
        speed=1.0,
        range=9.0,
        sight_range=9.5,
        radius=0.4,
        mass=2,
        is_air=False,
        target_type="all",
        elixir_cost=3,
        projectile_speed=10,
        projectile_asset="projectile_princess",
        projectile_trajectory="parabola",
        splash_radius=2.0,
    ),
    "princess_tower": EntityStats(
        id="princess_tower",
        name="Princess Tower",
        type="tower",
        hp=2534,
        damage=90,
        hit_speed=0.8,
        load_time=0.5,
        speed=0,
        range=7.5,
        sight_range=7.5,
        radius=1.5,
        mass=inf,
        is_air=False,
        target_type="all",
        projectile_speed=10,
        projectile_asset="projectile_arrow_basic",
        projectile_trajectory="parabola",
    ),
    "king_tower": EntityStats(
        id="king_tower",
        name="King Tower",
        type="tower",
        hp=4008,
        damage=90,
        hit_speed=1.0,
        load_time=0.2,
        speed=0,
        range=7.0,
        sight_range=7.0,
        radius=2.0,
        mass=inf,
        is_air=False,
        target_type="all",
        projectile_speed=15,
        projectile_asset="projectile_cannonball_large",
        projectile_trajectory="line",
    ),
}
